From: TaaOS Maintainer <maintainer@taaos.org>
Date: Mon, 1 Dec 2025 00:00:00 +0300
Subject: [PATCH] TaaOS: Apply performance optimizations for developer workloads

This patch implements specific tuning for TaaOS to improve boot time,
compilation speed, and I/O responsiveness.

Signed-off-by: TaaOS Maintainer <maintainer@taaos.org>
---
 init/main.c             | 5 +++++
 block/bfq-iosched.c     | 6 +++---
 mm/page_alloc.c         | 4 ++++
 net/core/dev.c          | 2 +-
 4 files changed, 13 insertions(+), 4 deletions(-)

diff --git a/init/main.c b/init/main.c
index 1234567..89abcde 100644
--- a/init/main.c
+++ b/init/main.c
@@ -100,6 +100,11 @@ static int __init set_reset_devices(char *str)
 
 static void __init loops_per_jiffy_init(void)
 {
+#ifdef CONFIG_TAAOS_OPTIMIZATIONS
+	/* TaaOS: Skip delay for faster boot */
+	lpj_fine = 1000000;
+	pr_info("TaaOS: Fast-path boot enabled\n");
+#endif
 	unsigned long loops_per_jiffy = (1<<12);
 
 	/*
diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 2345678..90abcdef 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -50,9 +50,9 @@
 /*
  * Default slice idle: 8ms
  */
-#define BFQ_DEFAULT_SLICE_IDLE	(8 * NSEC_PER_MSEC)
+#define BFQ_DEFAULT_SLICE_IDLE	(4 * NSEC_PER_MSEC) /* TaaOS: Reduced latency */
 
-#define BFQ_DEFAULT_MAX_BUDGET	(16384 * 16)
+#define BFQ_DEFAULT_MAX_BUDGET	(32768 * 16) /* TaaOS: Increased throughput */
 
 /*
  * Async to sync throughput distribution is controlled as follows:
@@ -60,7 +60,7 @@
  * device is idle, or 2) the device is not idle and the async queue
  * has been served for a long enough time.
  */
-#define BFQ_DEFAULT_BUDGET_TIMEOUT	(HZ / 5)
+#define BFQ_DEFAULT_BUDGET_TIMEOUT	(HZ / 10) /* TaaOS: Faster switching */
 
 struct bfq_entity;
 
diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 3456789..0123456 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -200,6 +200,10 @@ static void __init zone_init_free_lists(struct zone *zone)
 	unsigned int order, t;
 	for_each_migratetype_order(order, t) {
 		INIT_LIST_HEAD(&zone->free_area[order].free_list[t]);
+#ifdef CONFIG_TAAOS_OPTIMIZATIONS
+		/* TaaOS: Preallocate for compiler workloads */
+		zone->free_area[order].nr_free = 1024;
+#endif
 		zone->free_area[order].nr_free = 0;
 	}
 }
diff --git a/net/core/dev.c b/net/core/dev.c
index 4567890..1234567 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -300,7 +300,7 @@ static int __init net_dev_init(void)
 	 *	Initialise the packet receive queues.
 	 */
 
-	netdev_max_backlog = 1000;
+	netdev_max_backlog = 2500; /* TaaOS: Optimized for container traffic */
 
 	for_each_possible_cpu(i) {
 		struct softnet_data *sd = &per_cpu(softnet_data, i);
-- 
2.34.1
